### DISCUSSION TITLE
[Android] Enabling dataflow through OS triggered callbacks (ServiceConnection)

### DISCUSSION THREAD

**Discussion by szwl** [github/codeql#17097] [Q&A]
Hi all,
I'm trying to figure out a problem with taint propagation in an Android application.

**Example code:**
```java
  public static void sourceMethod(Context context, SomeObject options) {
    ServiceConnection connection =
        new ServiceConnection() {
          @Override
          public void onServiceConnected(ComponentName name, IBinder service) {
            try {
              ISomeService serviceInterface =
                ISomeService.Stub.asInterface(service);
                  serviceInterface.doStuff(options);
            } catch (RemoteException e) {} 
          }

          @Override
          public void onServiceDisconnected(ComponentName name) {}
        };

    Intent serviceIntent =
        new Intent().setClassName(context, "Some class name");

    context.bindService(serviceIntent, connection, BIND_AUTO_CREATE);
  }
```

**My query:**
```
module TaintPropagationTestConfig implements DataFlow::ConfigSig {
  predicate isSource(DataFlow::Node source) { 
      exists(MethodCall mc |
        mc.getMethod().hasName("sourceMethod") and mc.getMethod().isStatic() and source.asExpr() = mc.getArgument(1)
      )
  }

  predicate isSink(DataFlow::Node sink) { 
      // to debug taint propagation
      any()
  }
}


module TaintPropagationTestConfigFlow = TaintTracking::Global<TaintPropagationTestConfig>;

from TaintPropagationTestConfigFlow::PathNode source, TaintPropagationTestConfigFlow::PathNode sink
where TaintPropagationTestConfigFlow::flowPath(source, sink)
select sink.getNode(), sink, source, source.getNode()
```

**The problem:**

It seems that codeql doesn't analyze onServiceConnected method, so I'm unable to propagate taint to serviceInterface.doStuff.
I can workaround this by adding an `AdditionalTaintStep` eg.

```
private class ServiceConnectionStep extends AdditionalTaintStep {
  override predicate step(DataFlow::Node n1, DataFlow::Node n2){
    exists(VarAccess va | va.getVariable() = n1.asParameter().(Variable) and va = n2.asExpr())
  }
}
```

however, it seems to be a hack rather than a solution that would work for more applications.

**Questions:**
1. How can I instruct codeql that context.bindService will eventually trigger connection.onServiceConnected?
2. Is [AdditionalDataFlowNode](https://codeql.github.com/codeql-standard-libraries/java/semmle/code/java/dataflow/FlowSteps.qll/type.FlowSteps$AdditionalDataFlowNode.html) a solution here? If yes, how am I supposed to use it?

**Comment by jketema** [comment]
Hi,

Extending `AdditionalTaintStep` to handle this is a good approach. The same is, for example, done in the Java CodeQL library for Android's `AsyncTask::execute`

https://github.com/github/codeql/blob/c989e01197ee559e52b50bbd3466fc83da63be0e/java/ql/lib/semmle/code/java/frameworks/android/AsyncTask.qll#L16-L25

**Reply by szwl** [reply]
Thank you for the pointer, however I'm unsure if this approach would work here. In case of async task, the taint is propagated from `AsyncTask::execute` and `AsyncTask::executeOnExecutor` parameter to `AsyncTask::doInBackground` parameter.

In my case, `options` come from outer context and get used in anonymous implementation of `ServiceConnection`. There is no parameter passed to `onServiceConnected` that is relevant to the taint flow. Codeql seems to ignore implementation of `onServiceConnected` as it's never called - my goal is to notify Codeql that `bindService` in fact calls `onServiceConnected`.

**Reply by jketema** [reply]
Apologies. I had missed that you were interested in the `options` parameter. What you have is pretty much the best you can do, as `options` does not flow into `bindService` and subsequently to `onServiceConnected`.

**Reply by szwl** [reply]
Thanks - is there maybe a way to fake dataflow nodes to direct codeql in the right path? Eg. [AdditionalDataFlowNode](https://codeql.github.com/codeql-standard-libraries/java/semmle/code/java/dataflow/FlowSteps.qll/type.FlowSteps$AdditionalDataFlowNode.html)? It seems like what I need, but have hard time figuring out how to use it...

**Reply by jketema** [reply]
I'm not sure why you think that AdditionalDataFlowNode would help? You will still need to introduce addition steps to and from that additional node. Presumably a step from your options parameter to the node, and a step from the node to the use of options in `onServiceConnected`.

**Reply by szwl** [reply]
I'm not very good with CodeQL, so maybe indeed this is completely wrong :) It's challenging to me to understand parts of documentation (especially if there is no example using that API). 

My general idea would be to introduce another node whenever bindService call is detected - this node would be a fake method call on connection object (`connection.onServiceConnected`). My hope was that this way CodeQL wouldn't consider that path as unreachable and correctly taint `options` variable access. 

Access of variables from outer scope in lambda/anonymous functions are generally tainted correctly I checked with Java Supplier API that has similar properties to my case:

```java
// propagates taint correctly, finds doStuff sink
((Supplier<Void>) () -> {
  serviceInterface.doStuff(options);
}).get();

// doesn't find the doStuff sink, taint not passed 
Supplier<Void> ignored = ((Supplier<Void>) () -> {
  serviceInterface.doStuff(options);
});
```
There is a difference for CodeQL whether `get()` is or isn't called - I believe that finding what the difference is would help me solve my issue.

**Reply by jketema** [reply]
> I'm not very good with CodeQL, so maybe indeed this is completely wrong :) It's challenging to me to understand parts of documentation (especially if there is no example using that API).

The dataflow library is not the easiest piece of QL code, so that's fully understandable. Note that there's one example that uses `AddinitionalDataFlowNode`, see the PR that introduced the class: https://github.com/github/codeql/pull/14297.

The difference between your two `Supplier<Void>` examples strikes me as odd. You might want to open an issue in this repo (not a discussion) to flag that up. I would expect there to be no difference there. This is of course still very much different from the original issue, as all the code is known in this case, while it isn't in your original problem.

**Reply by szwl** [reply]
Thanks, I'll take a look at this PR.

I actually think this is working as intended - `ignored` supplier is never "`get()`", so its code would never run. However, understanding how is that distinction made, could be a key to solve my problem.

**Reply by jketema** [reply]
> Access of variables from outer scope in lambda/anonymous functions are generally tainted correctly I checked with Java Supplier API that has similar properties to my case:
> 
> ```java
> // propagates taint correctly, finds doStuff sink
> ((Supplier<Void>) () -> {
>   serviceInterface.doStuff(options);
> }).get();
> 
> // doesn't find the doStuff sink, taint not passed 
> Supplier<Void> ignored = ((Supplier<Void>) () -> {
>   serviceInterface.doStuff(options);
> });
> ```
> 
> There is a difference for CodeQL whether `get()` is or isn't called - I believe that finding what the difference is would help me solve my issue.

I'm not sure how similar this is to your situation, actually. Did you have a look at the actual dataflow path that is produced by the first of these two examples?

Note that the `get` method for `Supplier` is modelled here:

https://github.com/github/codeql/blob/ffab199ea84fc69b4dd2fd573eab912d04498584/java/ql/lib/ext/com.google.common.base.model.yml#L95
